File: DataStructures/Disjoint Set Union/atcoder/A66/A66.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

type UnionFind struct {
	parent []int
	rank   []int
}

func NewUnionFind(n int) *UnionFind {
	uf := &UnionFind{
		parent: make([]int, n+1),
		rank:   make([]int, n+1),
	}
	for i := 1; i <= n; i++ {
		uf.parent[i] = i
	}
	return uf
}

func (uf *UnionFind) find(x int) int {
	if uf.parent[x] != x {
		uf.parent[x] = uf.find(uf.parent[x]) // 経路圧縮
	}
	return uf.parent[x]
}

func (uf *UnionFind) union(x, y int) {
	rootX := uf.find(x)
	rootY := uf.find(y)
	if rootX == rootY {
		return
	}
	if uf.rank[rootX] < uf.rank[rootY] {
		uf.parent[rootX] = rootY
	} else {
		uf.parent[rootY] = rootX
		if uf.rank[rootX] == uf.rank[rootY] {
			uf.rank[rootX]++
		}
	}
}

func (uf *UnionFind) same(x, y int) bool {
	return uf.find(x) == uf.find(y)
}

func main() {
	scanner := bufio.NewScanner(os.Stdin)
	scanner.Scan()
	parts := strings.Split(scanner.Text(), " ")
	N, _ := strconv.Atoi(parts[0])
	Q, _ := strconv.Atoi(parts[1])

	uf := NewUnionFind(N)
	var output []string

	for i := 0; i < Q; i++ {
		scanner.Scan()
		query := strings.Split(scanner.Text(), " ")
		t, _ := strconv.Atoi(query[0])
		u, _ := strconv.Atoi(query[1])
		v, _ := strconv.Atoi(query[2])

		if t == 1 {
			uf.union(u, v)
		} else {
			if uf.same(u, v) {
				output = append(output, "Yes")
			} else {
				output = append(output, "No")
			}
		}
	}

	fmt.Println(strings.Join(output, "\n"))
}

File: DataStructures/Disjoint Set Union/atcoder/A66/A66.ts
import * as fs from 'fs';

const input = fs.readFileSync(0, 'utf8').trim().split('\n');
const [N, Q] = input[0].split(' ').map(Number);

// Union-Find データ構造
class UnionFind {
    parent: number[];
    rank: number[];

    constructor(size: number) {
        this.parent = Array.from({ length: size + 1 }, (_, i) => i);
        this.rank = Array(size + 1).fill(0);
    }

    find(x: number): number {
        if (this.parent[x] !== x) {
            this.parent[x] = this.find(this.parent[x]);
        }
        return this.parent[x];
    }

    union(x: number, y: number): void {
        const rootX = this.find(x);
        const rootY = this.find(y);
        if (rootX === rootY) return;

        if (this.rank[rootX] < this.rank[rootY]) {
            this.parent[rootX] = rootY;
        } else {
            this.parent[rootY] = rootX;
            if (this.rank[rootX] === this.rank[rootY]) {
                this.rank[rootX]++;
            }
        }
    }

    same(x: number, y: number): boolean {
        return this.find(x) === this.find(y);
    }
}

const uf = new UnionFind(N);
const output: string[] = [];

for (let i = 1; i <= Q; i++) {
    const [t, u, v] = input[i].split(' ').map(Number);
    if (t === 1) {
        uf.union(u, v);
    } else {
        output.push(uf.same(u, v) ? 'Yes' : 'No');
    }
}

console.log(output.join('\n'));
